### PR TITLE
Application Gateway: Rewrite Rules

### LINKED ISSUE #3284: Application Gateway: Rewrite rules
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Add Rewrite HTTP headers feature. I'll be sending a PR soon.

### New or Affected Resource(s)

azurerm_application_gateway

### Potential Terraform Configuration

```hcl
rewrite_rule {
  name = "rwrule"
  rule_sequence = 1
  condition {
    variable = "Location"
    pattern = "(https:?):\/\/.*azurewebsites\.net(.*)$"
  }
  response_header_configuration {
    header_name = "Location"
    header_value = "{http_resp_Location_1}://contoso.com{http_resp_Location_2}" 
  }
}
```

### References

* https://docs.microsoft.com/en-us/azure/application-gateway/rewrite-http-headers
* https://docs.microsoft.com/en-us/cli/azure/network/application-gateway/rewrite-rule?view=azure-cli-latest#az-network-application-gateway-rewrite-rule-create
* https://godoc.org/github.com/Azure/azure-sdk-for-go/services/network/mgmt/2018-12-01/network#ApplicationGatewayRewriteRule

### PR DESCRIPTION
**Pull request by mcharriere**
Adds support for rewrite rules #3284 

(fixes #3284)

### CODE REVIEW

File: azurerm/resource_arm_application_gateway_test.go [hunk 1]
```diff
@@ -3080,3 +3130,227 @@ resource "azurerm_application_gateway" "test" {
...
+      name = "${local.rewrite_rule_name}"
+      rule_sequence = 1
+
+      condition {
+        variable = "var_http_status"
+	pattern = "502"
+      }
+
+      request_header_configuration {
+	header_name = "X-custom"
```
Comment (katbyte) [inline azurerm/resource_arm_application_gateway_test.go]: Could we fix the formatting of this test?

File: azurerm/resource_arm_application_gateway_test.go [hunk 2]
```diff
@@ -3080,3 +3130,227 @@ resource "azurerm_application_gateway" "test" {
...
```
Comment (katbyte) [inline azurerm/resource_arm_application_gateway_test.go]: Could we fix the formatting of this test?

File: azurerm/resource_arm_application_gateway.go [hunk 3]
```diff
@@ -2283,13 +2429,168 @@ func flattenApplicationGatewayRequestRoutingRules(input *[]network.ApplicationGa
...
+							responseConfig := map[string]interface{}{
+								"header_name":  *config.HeaderName,
+								"header_value": *config.HeaderValue,
+							}
+							responseConfigs = append(responseConfigs, responseConfig)
+						}
+
+						ruleOutput := map[string]interface{}{
+							"name":                          *rule.Name,
+							"rule_sequence":                 *rule.RuleSequence,
```
Comment (katbyte) [inline azurerm/resource_arm_application_gateway.go]: We will need to nili check this

File: azurerm/resource_arm_application_gateway.go [hunk 4]
```diff
@@ -2283,13 +2429,168 @@ func flattenApplicationGatewayRequestRoutingRules(input *[]network.ApplicationGa
...
+				output["id"] = *config.ID
+			}
+
+			if config.Name != nil {
+				output["name"] = *config.Name
+			}
+
+			if rulesConfig := props.RewriteRules; rulesConfig != nil {
+				rules := make([]interface{}, 0)
+				for _, rule := range *rulesConfig {
```
Comment (katbyte) [inline azurerm/resource_arm_application_gateway.go]: We will need to nili check this

File: azurerm/resource_arm_application_gateway.go [hunk 5]
```diff
@@ -2283,13 +2429,168 @@ func flattenApplicationGatewayRequestRoutingRules(input *[]network.ApplicationGa
...
+					if rule.Name != nil {
+
+						conditions := make([]interface{}, 0)
+						for _, config := range *rule.Conditions {
```
Comment (katbyte) [inline azurerm/resource_arm_application_gateway.go]: We will need to nili check this

File: azurerm/resource_arm_application_gateway.go [hunk 6]
```diff
@@ -2283,13 +2429,168 @@ func flattenApplicationGatewayRequestRoutingRules(input *[]network.ApplicationGa
...
+							condition := map[string]interface{}{
+								"variable":    *config.Variable,
```
Comment (katbyte) [inline azurerm/resource_arm_application_gateway.go]: We will need to nili check this, and all of derefs here

File: azurerm/resource_arm_application_gateway.go [hunk 7]
```diff
@@ -2283,13 +2429,168 @@ func flattenApplicationGatewayRequestRoutingRules(input *[]network.ApplicationGa
...
+								"pattern":     *config.Pattern,
+								"ignore_case": *config.IgnoreCase,
+								"negate":      *config.Negate,
+							}
+							conditions = append(conditions, condition)
+						}
+
+						requestConfigs := make([]interface{}, 0)
+						for _, config := range *rule.ActionSet.RequestHeaderConfigurations {
```
Comment (katbyte) [inline azurerm/resource_arm_application_gateway.go]: and here

File: azurerm/resource_arm_application_gateway.go [hunk 8]
```diff
@@ -2283,13 +2429,168 @@ func flattenApplicationGatewayRequestRoutingRules(input *[]network.ApplicationGa
...
+							requestConfig := map[string]interface{}{
+								"header_name":  *config.HeaderName,
+								"header_value": *config.HeaderValue,
+							}
+							requestConfigs = append(requestConfigs, requestConfig)
+						}
+
+						responseConfigs := make([]interface{}, 0)
+						for _, config := range *rule.ActionSet.ResponseHeaderConfigurations {
```
Comment (katbyte) [inline azurerm/resource_arm_application_gateway.go]: and here

**Comment by mcharriere** [comment]
@katbyte I've made the changes in the flatten function checking for nil. Travis is failing but I think it's not related with the code.
I've also added a validation for the rule_sequence param.

Thanks for the review!!

**Comment by katbyte** [comment]
image

**Pull request merged**